### DISCUSSION TITLE
more app.jvm.cli options

### DISCUSSION THREAD

**Discussion by ennerf** [hydraulic-software/conveyor#15] [Ideas]
It'd be great if the extra cli executables could

**1) set a custom icon**
It's a bit odd for CLI utilities to use the same icon as a GUI application. It'd be great if it would default to the OS-console icon and could be manually specified if needed.

**2) disable the console window on Windows**
We have a few sub-UIs that need to be bundled and get launched by the main application (e.g. [here](https://youtu.be/E9csKYFPwaM?t=478)). It's great that I can add those as CLI apps, but it'd be nice if they didn't show a console window when users click on the extra executables directly.

**Comment by mikehearn** [comment]
Right - the assumption is that apps have a single GUI entry point (for the start menu/bundle/launcher), and other entry points are CLI tools for the user to invoke by name so they have to be in console mode and on the path. 

I didn't actually see any CLI tools in your video, so it seems that the app itself is split into several different GUI programs? The issue is, that the install folder will be protected by Windows so users can't enter it directly with Explorer unless they elevate. Does the UX of your software assume users will navigate to the install directory and run sub-programs directly, or is what you really want here multiple GUI entry points? If so, how should that work on macOS where you get one icon per bundle? The reason Conveyor is set up this way is because on macOS it's quite painful to ship apps that are actually several apps in one go. The user ends up having to manage them independently, and the security/launch services infrastructure doesn't much like apps depending on each other.

**Reply by ennerf** [reply]
Fair enough. I'm still at a stage where I'm testing the app image or extracting the zip, so I haven't gotten to the full installation part yet. This might be less of an issue then.

The main GUI requires relatively little memory, but occasionally it needs to e.g. show a log viewer that loads >1GB files and keep them in memory. Plotting that much data can also freeze the UI thread for a noticeable stutter, but I don't want that to interfere with the low-latency plotting of live charts. (The "imitation device" launcher I linked to above also runs as a separate independent process.)

I'm currently solving this by launching the log-viewer using the bundled `java` with memory constraints that are appropriate for the log file size. Bundling `java` isn't a great option though, and I think it should be ok to bundle it as a CLI app with a high `-Xmx`.

The user typically never needs to execute those apps manually, although in some cases it'd be convenient.

**Reply by mikehearn** [reply]
OK. Sounds like what's needed here is actually a new feature of internal launchers. These would require explicitly setting their console mode, wouldn't be placed on the path, and on macOS wouldn't trigger update checks at startup.

**Reply by ennerf** [reply]
That sounds like a good feature. For some apps I might still opt to keep them on path to allow `logviewer <filename>` though. I'll keep experimenting for now and think more about it when I get to the full release.

**Reply by mikehearn** [reply]
Alright. For now it's quickest to add a console key to the launchers so you can get rid of the annoying popup box. Internal launchers might follow later. 

Your app looks fantastic, by the way. Very slick. I hope I'm never being hunted by one of those hexapods!

**Reply by ennerf** [reply]
Thanks!

After using it a bit more, an `internal` app that does not show up on the path would be great. We have some tools that are a bit dangerous and shouldn't be launched by users without sanity checks.

The icons also turned out to show up in the Windows search, so those would be neat to be settable as well.

```
app.jvm.cli.foo-cli {
    main-class = com.foobar.FooTool

    # Set an explicit class path. You normally never need this because the
    # default of *.jar is good enough. 
    class-path = "some-prefix-*.jar"

    # JVM options added to the global list.
    options = [ -Xmx500M ]

     # Icon (same as app, empty defaults to system icon)
    icon = []

     # Internal tools are placed in e.g. /bin/internal and don't show up on the path
    internal = true

    # Disable the console output on Windows
    console = false
}
```

**Reply by mikehearn** [reply]
For the icons, I think they won't appear in search anymore if we make this internal launcher feature, so what the exact icon is wouldn't necessarily matter - you'd only see them in Explorer. Still, it's not hard to provide specific control over what's used.

If they're separate processes though, and showing their own GUIs, wouldn't they need an icon for the taskbar and task switcher? Or is the icon you set via JavaFX sufficient there?

**Reply by ennerf** [reply]
Yes, the icon would have to be set before JVM launch to work  on at least macOS as the dock is only accessible via the `-Xdock:icon` args

**Reply by mikehearn** [reply]
The next release will have a per-launcher `console` key, as well as allowing JVM options to be controlled per launcher/machine combination. So for instance you can pass a JVM flag (and therefore set a system property) for only the Apple Silicon Mac version of a specific launcher.

**Comment by ennerf** [comment]
Another interesting behavior I've observed is that the icon for the primary application shows a white border

![image](https://user-images.githubusercontent.com/5491587/182800174-9f251257-17df-46be-b7eb-28cb86e44ee7.png)

while the same icon files for the CLI apps show correctly

![image](https://user-images.githubusercontent.com/5491587/182800313-76a07c1d-e92f-4232-b466-679a3d691281.png)

I originally thought this was an icon caching issue, but trying it on an otherwise blank VM yields the same result. Is there anything you're doing differently between the two icons? I'm currently adding `2^x` resolution files from `16x16` to `256x256`.

**Reply by mrbubble** [reply]
Is that Windows 11? The border is called "plating", and [Microsoft makes some decisions on when to display it or not that are a bit hard to understand](https://docs.microsoft.com/en-us/windows/apps/design/style/app-icons-and-logos#unplated-assets). There is a way to override it with icon metadata which we are planning to support in a future Conveyor release. For now, you can try changing the background color with undocumented key: `app.windows.manifests.msix.background-color`, which maps to the [uap:VisualElements](https://docs.microsoft.com/en-us/uwp/schemas/appxpackage/uapmanifestschema/element-uap-visualelements) BackgroundColor attribute. At least that gives you some control of what gets displayed until we make a release that lets you get rid of plating entirely.

**Reply by ennerf** [reply]
No, it's windows 10. The CLI icons show up correctly, so I thought it might be an easier fix. Here is a full sample project with the same behavior https://github.com/ennerf/HdrHistogramVisualizer

I'm out today, but I'll try your suggestion when I'm back.

**Reply by ennerf** [reply]
On Windows 11 it shows a transparent border

<img width="494" alt="image" src="https://user-images.githubusercontent.com/5491587/183107659-626e44fc-2d77-4e6f-9532-4417608e5670.png">

**Reply by ennerf** [reply]
The `app.windows.manifests.msix.background-color` works, thanks!
